### ISSUE TITLE
GPflow1 to GPflow2: more Cholesky decomposition problems

### ISSUE THREAD

**Issue by tdewolff** [GPflow/GPflow#1197]
GPflow: 2.0.0rc1
Tensorflow: 2.0.0

We're in the process of moving our code from GPflow1 to GPflow2, but while doing so it seems that Cholesky decomposition problems occur much more often than previously. One reason was that the likelihood variance in the GPR model became VERY small all of a sudden. This variance is added to the diagonala, but if it is smaller than 1e-12 for example Cholesky fails. In general, parameter estimation through the SciPy/Adam optimizers gives poorer results. Sometimes parameters become NaN while optimizing.

Has anything fundamentally changed between GPflow1 and GPflow2 that we need to be aware of? For example, the inner working of `gpflow.transforms.positive` is similar to the new `gpflow.utilities.positive()`? The latter seems to more easily result in NaN (when input is negative). Anything we need to be aware of?

PS: overall GPflow2 is a big improvement! Thanks for the work!

**Comment by awav** [comment]
@tdewolff, thanks a lot! 

In GPflow 1.0, positive values had a lower bound 1e-6. It was removed in favour of a more customisable solution. Now you can set a lower positive bound to a value which you think is most appropriate.

```python
assert gpflow.config.default_positive_minimum() is None  # by default it is `None`
gpflow.config.set_default_positive_minimum(1e-6)
```

https://github.com/GPflow/GPflow/blob/develop/gpflow/config/__config__.py#L153

You also can change the positive bijector type (by default it is `softplus`). You have two choices "exp" and "softplus".

```python
assert gpflow.config.default_positive_bijector() == "softplus"
gpflow.config.set_default_positive_bijector("exp")
```

https://github.com/GPflow/GPflow/blob/develop/gpflow/config/__config__.py#L143

**Comment by st--** [comment]
I actually just ran into a related issue where in my custom model I had a zero-initialized positive-constrained parameter, `p = gpflow.Parameter(0.0, transform=gpflow.utilities.positive())`. This is represented as `-inf` in the unconstrained space, which is fine in the forward pass (back to 0.0), but results in `nan`s when computing gradients. What can we do to improve the user experience around positive transforms ?

**Comment by tdewolff** [comment]
Thanks for the quick response! The `gpflow.config.set_default_positive_minimum` function solved most of our problems. For the `gpflow.config.set_default_positive_bijector` we'll have to wait for a new release I guess. I see the default mode for GPflow1 was `x_con = log(1+exp(x_uncon))`, this is different from the bijector `exp` I suppose? Not sure if this has a big impact in the way the parameter are trained.

Another issue I'd like to mention (also w.r.t. #1198) is that previously in GPflow1 it was fine to set a parameter to zero eventhough the positive minimum was `1e-6`. Right now, when setting the positive minimum using `gpflow.config.set_default_positive_minimum` to `1e-6`, putting a zero in `Parameter` (or using `assign`) causes a `nan`. Would it be beneficial if both the `Parameter.__init__` and `Parameter.assign` do a lower bound check to ensure it is bigger than or equal to `default_positive_minimum`?

**Comment by st--** [comment]
@tdewolff I would suggest using the `develop` branch instead of rc1, it's got a lot more bugfixes.

The default "positive" bijector in both gpflow1 and gpflow2 is 'softplus' (y = log(1+exp(x)); you are right, 'exp' which is just y = exp(x) is different. They will have different optimisation behaviour and depending on your problem specifics one or the other may optimise better/faster, though you should get pretty much the same answer upon convergence.

In GPflow1 it *looked* like you could set a "positive"-constrained parameter to zero. In fact, under the hood it was simply set to the lower bound, i.e. 1e-6, even if it looked like it was zero (all that complexity with python representation and session representation etc. - just be glad it all went away). #1198 specifically introduces a validation check as you say! It will prevent you from setting a parameter with constraint to any value outside the range. I.e. `gpflow.Parameter(0.0, transform=positive())` will then throw an error. With a lower bound, valid parameter values are all values strictly greater than the lower bound. I.e. `Parameter(1e-6, transform=positive(lower=1e-6))` would error, `Parameter(1.1e-6, transform=positive(lower=1e-6))` would work fine.

Does that answer your questions/issues? If not please clarify!

**Comment by tdewolff** [comment]
We're developing a library that builds on top of GPflow, I'm afraid we have to depend on a specific release/version of GPflow for our dependency management. As we're not yet due for release we're fine with waiting for an official v2 for now. It's a release candidate so it's totally understandable that there are many bugfixes on their way!

What I wanted to say about the parameter lower bound is that in GPflow1 it was fine to set a parameter to zero even though it would be set to the lower bound internally, that lower bound would be our effective 'zero' value which was totally fine. Right now setting a value lower than the lower bound effectively sets the parameter to `nan` which causes a non-trivial error to debug. With the new PR this should become a helpful error message which is great! But in our usecases we oftentimes just want to set the parameter to be 'effectively' zero, be it zero or the lower bound. I've added some wrapper code to clamp lower values to the lower bound which is a bit clumsy. I suppose clamping inside GPflow2 would avoid surprises for users coming from GPflow1, but perhaps an error would be a better solution ;-). Thanks for the help!

**Comment by st--** [comment]
@tdewolff your comment is effectively what @jameshensman is suggesting as point 4 - if you've got any ideas for how to actually implement that, please show! Even if it's just the "clumsy wrapper code", I think if you shared that it might be helpful for a better understanding and/or other users!

Thoughts not specifically relevant to this issue, but because you mention it and I don't know how else to reply to you:
1) At least while you're developing the gpflow2-compatible version of your library (is it this one: https://github.com/GAMES-UChile/MultiOutputGP-Toolkit ?), could you not just use the develop branch? `pip` can easily install from any branch on a git repository, not just release tags. Could it make your life easier to stay on top of our develop while things are still changing around?
2) GPflow2 isn't "set in stone" - if in your work with & on top of gpflow you identify any pain points / issues with what it does (or doesn't!) do - regardless of whether that's something that's carried over from GPflow1 or new to gpflow2 - we would very much appreciate your thoughts / comments / feedbacks! We can only make your experience smoother if you tell us about the bumps :)
3) In 2) I said "we" and "you" but actually we would love more people to contribute to GPflow - so if you've got any features you think might actually fit better in the base package, or otherwise want to help improve what we've got, please join us! There is also a slack workspace (gpflow.slack.com) that is a bit underutilised at the moment but that we want to make more use of. Help wanted. :-)
